sklearn/ensemble/_gradient_boosting.pyx

```diff
@@ -84,22 +88,111 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
         shape ``(n_samples, K)``.
     """
     cdef Py_ssize_t i
-    cdef int32 node_id
     cdef Node *node
     for i in range(n_samples):
         node = root_node
         # While node not a leaf
-        while node.left_child != -1 and node.right_child != -1:
+        while node.left_child != TREE_LEAF:
             if X[i * n_features + node.feature] <= node.threshold:
                 node = root_node + node.left_child
             else:
                 node = root_node + node.right_child
         out[i * K + k] += scale * value[node - root_node]
 
+def _predict_regression_tree_stages_sparse(np.ndarray[object, ndim=2] estimators,
+                                           object X, double scale,
+                                           np.ndarray[float64, ndim=2] out):
+    """Predicts output for regression tree inplace and adds scaled value to ``out[i, k]``.
+
+    The function assumes that the ndarray that wraps ``X`` is csr_matrix.
+    """
+    cdef DTYPE_t* X_data = <DTYPE_t*>(
+        <np.ndarray[ndim=1, dtype=DTYPE_t]>X.data).data
+    cdef INT32_t* X_indices = <INT32_t*>(
+        <np.ndarray[ndim=1, dtype=INT32_t]> X.indices).data
+    cdef INT32_t* X_indptr = <INT32_t*>(
+        <np.ndarray[ndim=1, dtype=INT32_t]> X.indptr).data
+
+    cdef SIZE_t n_samples = X.shape[0]
+    cdef SIZE_t n_features = X.shape[1]
+    cdef SIZE_t n_stages = estimators.shape[0]
+    cdef SIZE_t n_outputs = estimators.shape[1]
+
+    # Initialize output
+    cdef float64* out_ptr = <float64*> out.data
+
+    # Indices and temporary variables
+    cdef SIZE_t sample_i = 0
+    cdef SIZE_t feature_i = 0
+    cdef SIZE_t stage_i = 0
+    cdef SIZE_t output_i = 0
```

Comment:
Again, I didn't check performance gain, but here:
https://cython.readthedocs.io/en/latest/src/quickstart/cythonize.html#typing-variables
It's said:
>Since the iterator variable i is typed with C semantics, the for-loop will be compiled to pure C code.

This is why I predefined them.

And additionaly, we have same variables in _tree.pyx.